### ISSUE TITLE
[2D MVP] Milestone 3: Snapping in 2d viewer

### ISSUE THREAD

**Issue by sanmont3drepo** [3drepo/3drepo.io#4941]
### Description
For setting the 2d vector for calibration we want to provide the user with snapping for convenience. 

### Goals
- [x] As a user I want to be able to snap to lines in the 2d viewer
- [x] Snapping should occur as the cursor moves
- [x] Cursor should snap to Corners, Intersections and Curves/Lines
- [x] There is a method that can be called externally that will return the snapped point
- [x] It is possible to enable or disable snapping

### Tasks
- [x] Add skeleton infrastructure for testing snapping behaviour
- [x] Test polyfil to process paths into line segments
- [x] Make functions to translate between container rect and svg coordinate system for svgImage
- [x] Make cursor for debugging
- [x] Find out which [SVG elements](https://developer.mozilla.org/en-US/docs/Web/SVG) need to be supported
- [x] Add support for `polyline`
- [x] Add support for `path`
- [x] Choose a suitable acceleration structure for Line intersection tests
- [x] Get some more SVGs from ODA for testing
- [x] Extract node-points from primitives for closest point tests and add to collector
- [x] Add logic to compute intersections and add to collector
- ~Add acceleration structure for points~
- [x] Integrate point tests into snapping logic and add type to results structure
- [x] Add check to make sure svg element has width and height properly set (otherwise the mapping breaks)
- [x] Make cursor update on zoom
-  ~PolyCollector should not add duplicate points~
- [x] Add Bezier Curve closest point test
- [x] Add Curve & Line intersection test
- [x] Add Curve & Curve intersection test
- [x] Add curve self-intersection test
- [x] Add rect support
- [x] Add circle support
- [x] Add ellipse support
- [x] Add line support
- [x] Add polyline support
- [x] Add polygon support
- [x] Add path support
- [x] Optimise the intersection tests (maybe)

**Comment by sebjf** [comment]
A few thoughts on this...

It seems there are a few approaches:

1. Image Space

This is what is used in the broadphase in the 3D viewer and involves applying an image feature detector to a patch from the Canvas. 

This is fast and we know it works from the 3D viewer. 

The main issue is that it cannot select the centre of lines or intersections (it can only select edges of the stroke, and corners of stroke interactions), and the precision is limited to whatever the rasterization resolution is, which depends on the zoom level.

In the 3D viewer anything snapped via this method is then refined using geometry tests.


2. DOM with [`getPointAtLength`](https://developer.mozilla.org/en-US/docs/Web/API/SVGGeometryElement/getPointAtLength)

This method will return points in the viewport for any `SVGGeometryElement`, which covers almost everything except reference items (embedded images).

This is the only DOM method currently support by all browsers.

The issue with this approach is that we get very little information about the underlying primitive. The only option is to sample all paths to an arbitrary precision, and perform tests on these segments.

This could end up very expensive, especially if we ever need to rebuild the set (e.g. to make higher precision one on zooming).


3. DOM with `getPathData` polyfill

The SVG Working Group's proposed [`getPathData`](https://svgwg.org/specs/paths/#InterfaceSVGPathData) with normalisation, will provide a minimal set of 2D primitives that it would be practical to write a series of tests against.

This is about as efficient as it is possible to get. The segments can be arranged into a bounding hierarchy for quick closest-point tests and intersection tests. The only primitive that would be computationally expensive to deal with would be the cubic beziers.

The problem is that this API is not supported by Chromium, and indeed Chromium decommissioned the alternative API to this almost 10 years ago, and hasn't seen fit to introduce the new one!

There is a polyfill available: https://github.com/jarek-foksa/path-data-polyfill, which works by effectively parsing bits of the SVG itself.

Relying on a polyfill that does its own parsing isn't great. Though one compensatory factor is that most of the SVGs we will handle will be produced by bouncer, so if we can make sure it works with whatever's output by ODA we should cover most use cases.


4. WASM based parser

The nuclear option is to get a reliable SVG parser in another language (e.g. the one from the Chromium source), build our snapping services around it, and cross-compile it into WASM.

This allows us to avoid the spotty support from browser manufacturers, and will give as good performance as we can get other than relying on user-agent implementations.



Other things to consider are that DOMParser isn't available in Web Workers, so for 2 & 3 we need to parse the SVG in the main thread (though I believe it's then possible to hand the graph to a web worker, but am not sure).

**Comment by sebjf** [comment]
A brief update on this...

I have implemented the Harris Detector, which we chose for the 3D viewer.

It works well as a broadphase for SVGs. For example, here is the feature map configured to show Edges:

![image](https://github.com/user-attachments/assets/826d487f-0375-441a-8b09-547165416305)

It also works with PNGs, at roughly the native resolution:

![image](https://github.com/user-attachments/assets/ad52ecc8-82de-4182-a504-72620b6c38cc)

It does *not* work with blurry rasterisations of PNGs however. This is expected because the bilinear interpolation is designed to smooth out edges. We need to look at changing the sampling of PNGs, so that when zoomed in we perhaps use point-based sampling.

![image](https://github.com/user-attachments/assets/73f1d949-aa63-4af1-8990-20238a6bfefb)

Importantly, it also does *not* work with single pixel lines...

![image](https://github.com/user-attachments/assets/7e1e8fbf-ffd7-4220-931a-2f5f82fdfe11)

This is expected because the Harris Detector is designed to detect the edge of *regions*. 

The mechanics behind it are that the first step is to derive the image gradients, which we do with a Sobel operator. However, when running over a line 1 pixel wide, each side of the kernel of this operator will cancel the other out.

(Zooming in on a line we see it detected as a region with two edges...)

![image](https://github.com/user-attachments/assets/f7160f0e-a2cf-4ac3-a0c0-bd9ce8555805)

We can either try to get an image gradient that is sensitive to lines, or run a second detector dedicated to lines.

**Comment by AsiteZane** [comment]
Agreed, we shouldn't do it for raster images.

**Comment by sebjf** [comment]
Brief update on this,

![image](https://github.com/user-attachments/assets/9bee2877-a618-4cae-9671-2e01ce2405a4)

The [polyfill](https://github.com/jarek-foksa/path-data-polyfill) appears to work on at least some images. Above, the lines (only lines, not curves) extracted from the svg are rendered to a separate canvas. This canvas is scaled (not the primitives), so is in the svg-space. We can see the lines are being extracted correctly, and we can map cursor positions between the viewer and the svg's local space.

We next need to determine a suitable acceleration structure.

**Snapping Goals**

Based on Revit, the features we'd like to identify are:

1. Endpoints of Lines and Curves
2. Intersection points of Lines and Curves
3. Closest point to lines

We are not interested in Text, or arrowheads.

**Acceleration Structures**

The user can zoom to arbitrary degrees, so cursor sensitivity defined in pixels will change in svg space dramatically (unlike for the 3d viewer where its all done in screen space). 

_BVH_

A bounding volume hierarchy is a known quantity as we use it in number of places. BVHs can have highly efficient build times that do not depend on insertion order, depending on the construction method (one of the only structures for which this is true).

The advantages of BVHs (assuming AABBs - in this case we'd implement an [R-Tree](https://en.wikipedia.org/wiki/R-tree)) are that the intersection tests are going to be very, very efficient in 2D. It can also be quite shallow if the number of nodes per level is relatively high (e.g. 4-8), which will help with memory overheads.

The main advantage is that it can handle curves more easily than other methods such as the Quadtree, as we only need to determine the maximum bounds for a curve.

BVHs can be searched at arbitrary distances using a Minkoswki sum depending on zoom level easily. They can also be self-intersected to find intersection points - although we would want to do this only once and store the intersections in another hierarchy.

Though the intersection tests themselves are going to be highly efficient, the tree is not as an efficient partition as something like a quadtree, due to the large potential overlaps between primitives leading a more expensive search for each query.


_Quadtree_

In [Quadtrees](https://www.cs.umd.edu/~hjs/pubs/SameCGA88b.pdf), space is regularly subdivided as needed to reach a condition that minimises the number of primitives at each node. (This could be a count of primitives per node, or a maximum depth.) When lines are considered, they are split into q-edges, which are the segments of the lines that sit within the node.

Searching the structure is fast, as we can search a region of a particular size simply by picking the correct level in the tree to start from.

What is particularly interesting about this structure, is that if two lines intersect anywhere, they will always do so at a leaf node that has them (their q-edges) in common. So, after building the tree we can just take the leaf nodes as a set of optimised pairs to look for intersections with.

The main complication with quadtrees are that we would have to figure out how to make them work with Cubic Beziers (CBs). The same principles that apply to lines (i.e. q-lines, and the termination criteria) should be the same, but we'd need to work out what a q-edge of a curve looks like in memory, for example.

The other issue with quadtrees is the possibly considerable memory consumption in JavaScript, though this is likely to be true of any structure that can't be implemented as an array (BVHs are the only structure that could be implemented as an array).

We also need to consider the building time of the tree - though such trees could also be precomputed and cached, either by the client in indexeddb, or perhaps even on .io.


_kd-tree_

It is worth mentioning kd-trees, though they won't be a solution for everything because they don't support closest distance tests for lines/curves.

kd-trees split different dimensions at different levels. For 2D, we'd not use actual kd-trees, but either use a BSP (binary space partitioning), or create effectively something like an irregular or point quadtree, where a split occurs on both dimensions at each point.

Searching 2d-trees is very fast, but the insertion order determines the depth, which is not ideal.


_Spatial Hash_

A spatial hash would only be useful for looking at closest points, however in this regard it is by far the fastest to search.

The problem with such a hash is that it has a fixed cell size, meaning when zooming out more cells would need to be checked, counteracting some of its advantages.

A spatial hash would need all the intersection points pre-defined, which means we'd need to use another structure to find the intersections.

**Comment by sebjf** [comment]
Continuing notes...

A proof of concept snapping implementation has been created using QuadTrees, that snaps a cursor in the image container to a line from the SVG.

There are many caveats to this:

1. It doesn't snap to points explicitly
2. It doesn't detect intersections
3. Its not optimised
4. The SVGs output by ODA don't work because the PolyLines are not picked up,
5. There are some slight offsets in the test code for the cursor
6. etc

However, the poc does fundamentally work, as can be seen...

![image](https://github.com/user-attachments/assets/bc792bd5-dd00-40f7-bd5f-6a36fa05ceae)

The next step is to clean up the test visualisations, so we can be sure the results are accurate when converting between the coordinate systems (i.e. we place a point, then zoom in and see it maintains accuracy greater than sampling resolution when it was placed).

After this, the most important thing is to profile the code to check its performance, as if we need to do something like build a WASM library in C, or rely on async snapping via WebWorkers, we need to know that now.

If we are confident that this implementation can scale, then we can add in end-point snapping, and begin to detect intersections.

From what we can see of the performance so far, with the map with 80,000 line primitives the quadtree is built and sampled very quickly; though sampling it should be noted is only occurring on clicks at the moment, not each mouse move. The tree depth limit is important. With a depth of 24 for example the tree building process becomes unusable. With a tree depth limit of 14 (what its set to currently), it builds almost instantly and results in just over half a million nodes with the largest leaf node having 25 lines.

**Comment by sebjf** [comment]
When testing with the carpark stress test, as 12 level quadtree takes about 5 seconds to produce. The intersection tests range from 1 ms to 30 ms, depending on the range.

This is because the quadtree with a max limit of 12 can end up with over 196 elements in the leaf nodes. A depth of 12 creates a little over 5 million nodes.

A binary r-tree by comparison would have about 1.6 million nodes across 19 levels.

Given the likelihood of encountering pathological svgs that mean the quadtree will top out the maximum depth in many cells, it is worth creating an r-tree implementation for comparison as well, as given the high leaf counts it may not take that much longer to traverse.

**Comment by sebjf** [comment]
With an [STR](https://ia800900.us.archive.org/27/items/nasa_techdoc_19970016975/19970016975.pdf) R-Tree implementation given 10 children per node, the tree takes about 1.5 seconds to build. The intersection tests range from 0 to 5 ms, depending on the range.

While in theory the quadtree should over better spatialisation, the STR - Sort-Tile-Recursive algorithm - offers a close approximation in 2D, without the extraneous nodes.

Behaving better in both build and query time then the R-Tree appears to be the best structure, certainly for the line-distance snap.

We will likely need further tests to determine if it is best for the points. Most likely it will still have best performance, but only when we have a separate tree constructed to make a best fit to the points. It is worth comparing it with a kd-tree however - the SRT building heuristic is not far off a kd-tree construction.

**Comment by sebjf** [comment]
Another update...

A kd-tree has been implemented for closest points, which can be queried for the Revit Sample SVG and most others in between 0 and 0.3 ms. For the carpark, it is typically the same, though some locations can take up to 5 ms.

This performance is as expected. There is an optimisation available, which is to store the exact point that the split is made on, at each node, instead of having only leaf & branch nodes. This may allow earlier termination of the traversal. However the current profiles do not suggest this is necessary.

The main issue is with the construction time of the tree, which is about 17 seconds for the car park. (For the Revit Sample it is about 141 ms.)

This may be too long in absolute terms. In the first instance we can look at possible optimisations in layout for the structure. Other options include delivering a pre-built object, or offloading to a web worker (though it would still be 17 seconds before snapping would "work").

The next steps are to extract intersection points from the RTree, and build some logic to swap the cursors etc based on the intersection point type.

**Comment by sebjf** [comment]
A helper class was constructed for building a predefined list of intersection points. However this immediately revealed some performance problems: the Revit Sample had over 2.9 million potential intersections, though this list took only a few ms to build.

The Carpark however failed to iterate all potential intersections at all. Simply counting the intersections took too long, and adding them to a list caused an overflow.

A better approach could be to drop the kd-tree entirely and perform both closest-edge, closest-node and closest-intersections using the R-Tree.

This would be done by tweaking the current traversal to return all leaf nodes that overlap the query circle. The closest-nodes would be extracted from the primitives in the same way as they'd be when building the point-acceleration structure. A second stage could then look for intersections as well.

**Comment by sebjf** [comment]
Counting simply the number of overlapping nodes, the Revit Sample has up to a couple hundred nodes at max extents. The carpark however has over 38,000 leaf nodes at maximum extents.

Depending on how many nodes the browser can effectively check and maintain real-time performance (its probably much higher than we expect), one approach could simply be to not snap if the number of nodes is above a certain threshold, as in this case there will be so many snapping points the user wouldn't be able to control it effectively regardless.

**Comment by sebjf** [comment]
Updating the RTree to handle all intersections, the node and edge queries are as fast as before, meaning the kdTree is superfluous.

For intersection tests, an NxN intersection check of overlapping leaf nodes runs below a few ms in most cases, but in the case of the car park at maximum extents, can result in noticable lag...

```
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 1919.6000000238419, numNodes: 19178, numIntersectingPairs: 9391760}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 97.30000001192093, numNodes: 5428, numIntersectingPairs: 424314}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 110.09999996423721, numNodes: 5599, numIntersectingPairs: 430918}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 108.39999997615814, numNodes: 5663, numIntersectingPairs: 439360}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 102.19999998807907, numNodes: 5832, numIntersectingPairs: 481451}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 360.2000000476837, numNodes: 8418, numIntersectingPairs: 2404347}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 101.5, numNodes: 5444, numIntersectingPairs: 432086}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 93.5, numNodes: 5340, numIntersectingPairs: 419875}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 95.09999996423721, numNodes: 5447, numIntersectingPairs: 420509}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 101.39999997615814, numNodes: 5724, numIntersectingPairs: 432318}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 63.80000001192093, numNodes: 4830, numIntersectingPairs: 323671}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 213.0999999642372, numNodes: 7002, numIntersectingPairs: 1439782}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 875.9000000357628, numNodes: 12975, numIntersectingPairs: 6126426}
IntersectionTestResults {closestEdge: Vector2, closestNode: Vector2, queryTime: 3642, numNodes: 25217, numIntersectingPairs: 19336215}
```

Times above are in ms.

`numIntersectingPairs` demonstrates that a a pre-pass is necessary (as otherwise, the number of pairs undergoing geometry tests would be much higher), but that the model still has a prohibitive number of actual primitives whose AABBs overlap. (These figures do not include actual intersection tests between primitives.)

![image](https://github.com/user-attachments/assets/821d184c-c07b-4e3f-a45e-354c9b67d583)

The figure above illustrates the zoom level and approximate search radius at that level. That region contains 22608 individual primitives/17 million potential intersections.

**Comment by sebjf** [comment]
With regards to the elements we need to support, it seems ODA predominantly uses the Path and Polyline parameters.

Still, the full list of SVGElement subclasses is as follows,

```
SVGElement
SVGSVGElement 
SVGCircleElement 
SVGClipPathElement 
SVGDefsElement 
SVGDescElement 
SVGEllipseElement 
SVGFEBlendElement 
SVGFEColorMatrixElement 
SVGFEComponentTransferElement 
SVGFECompositeElement 
SVGFEConvolveMatrixElement 
SVGFEDiffuseLightingElement 
SVGFEDisplacementMapElement 
SVGFEDistantLightElement 
SVGFEDropShadowElement 
SVGFEFloodElement 
SVGFEFuncAElement 
SVGFEFuncBElement 
SVGFEFuncGElement 
SVGFEFuncRElement 
SVGFEGaussianBlurElement 
SVGFEImageElement 
SVGFEMergeElement 
SVGFEMergeNodeElement 
SVGFEMorphologyElement 
SVGFEOffsetElement 
SVGFEPointLightElement 
SVGFESpecularLightingElement 
SVGFESpotLightElement 
SVGFETileElement 
SVGFETurbulenceElement 
SVGFilterElement 
SVGForeignObjectElement 
SVGGElement 
SVGImageElement 
SVGLineElement 
SVGLinearGradientElement 
SVGMarkerElement 
SVGMaskElement 
SVGMetadataElement 
SVGPathElement 
SVGPatternElement 
SVGPolygonElement 
SVGPolylineElement 
SVGRadialGradientElement 
SVGRectElement 
SVGStopElement 
SVGSwitchElement 
SVGSymbolElement 
SVGTextElement 
SVGTextPathElement 
SVGTSpanElement 
SVGUseElement 
SVGViewElement 
```

Out of these, the following are edge-like geometric types (i.e. they can stroke lines or curves, and have nodes that can be snapped to). This includes primitives like circles, but not area fills like gradients, for example.

This list corresponds to the [Basic Shapes](https://www.w3.org/TR/SVG11/shapes.html) defined by the SVG specification (and the SVGPathElement)

SVGRectElement
SVGCircleElement 
SVGEllipseElement
SVGLineElement
SVGPolylineElement
SVGPolygonElement
SVGPathElement


All the above shapes can be represented by an SVGPathElement and the functions to do so are defined in the SVG spec.

It may be worth however performing intersection tests against the primitives themselves, as these might be more efficient than against b-curves.



Honorary mentions include:

SVGDefsElement
SVGMarkerElement
SVGSymbolElement

These define elements or templates that are rferenced elsewhere. For example, the [SVGMarkerElement](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/marker) is like a def specifically for arrowheads and similar.

**Comment by sebjf** [comment]
The next step for this is to add support for Bezier Curves. SVG specifies that Bezier Curves are cubic. The three tests required are

1. Closest Point
2. Line-Cubic Intersection
3. Cubic-Cubic Intersection

Out of these, the most algorithmically complex is the Closest Point. 

The most [performant method](https://inria.hal.science/inria-00518379/document) expresses the distance as a [quintic polynomial](https://www.tinaja.com/glib/bezdist.pdf) and finds the roots. As there is no analytic method for finding the roots of a quintic (more on that later), an iterative method (Newton) is used. The intervals to searched are culled based on [Sturms theorem]()https://en.wikipedia.org/wiki/Sturm%27s_theorem). 

There are a number of numeric methods based on finding intervals on the curve itself (e.g. [here](https://www.sciencedirect.com/science/article/pii/S0377042708005785?via%3Dihub)) but they are all slower than operating with polynomial roots.

Line-Cubic Intersection and Cubic-Cubic Intersection are surprisingly simpler. This is because the equation of the intersection point can be [derived by substitution](https://math.stackexchange.com/questions/2347733/intersections-between-a-cubic-b%c3%a9zier-curve-and-a-line?rq=1), and we only need to solve for the curve or line "parameter". For the line case, this involves solving a cubic function.

For the curve-curve case, it is a quantic function.

Equations exist for solving polynomials of order 4 or below, but not above. However, naively writing them out will result in stability issues due to quantisation errors. The solvers must be carefully implemented. (For example, these described by [David Eberly](https://www.geometrictools.com/Documentation/LowDegreePolynomialRoots.pdf)).

Something worth considering before choosing an implementation, is that the Closest Point does not require the same level of precision as the intersection tests, owing to the fact the cursors precision will be dependent on the current zoom level.

There is a question of whether if we were to combine hueristics designed for cubic Beziers with some simpler numerical approach (Chen's method is designed for arbitrary order Beziers), the result might actually be more performant, so it is worth finding a few more algorithms to read first.

**Comment by sebjf** [comment]
Following on from the above, there is now an implementation of closest point on Bezier, which is the most complex test we will have to do for snapping. 

**Some notes are below for documentation purposes.**

The approach taken is the root-finding method, which seems to be the agreed upon optimal method. The idea behind this is to express the distance between a point and a curve as a polynomial. The equation that approximates this distance is given originally by [Zhou](https://link.springer.com/article/10.1007/BF01201903) as $\ (p-q(u)) \cdot q'(u) $.

Geometrically, this is the dot product of the vector between the query point _p_ and a point the curve at _u_, and the tangent of the curve. Since the dot product is 0 when the vectors are perpendicular, this evaluates to 0 when the shortest line is perpendicular to the curve: a necessary condition for $` q(u) `$ to be the closest point.

(This is all well introduced by [Chen et al.](https://ieeexplore.ieee.org/document/4392595))

Since this expanded equation consists of dot products, the resulting polynomial is univariate for _u_. 

([nictvg](https://computingandrecording.wordpress.com/2017/03/20/closest-point-to-a-cubic-spline/) has kindly done the first part of working out the actual coefficients from the control & query points. The equation can also be split into two such that we can compute half the coefficients and store them for future tests ($`-q(u) \cdot q'(u) `$), then when we update the query point we have only three dot products to recompute $` p \cdot q'(u) `$.)

Below is a plot of a Bezier with varying query points (red circles).

![image](https://github.com/user-attachments/assets/aec9eb79-114d-4110-a7e1-825ddcff5484)

The 'top' line is the distance between the curve and the query point, evaluated numerically for various points along _u_. The 'bottom' line is $` (p-q(u)) \cdot q'(u) `$ (also evaluated numerically, at the same points).

We can see from the above how this equation approximates the derivate of the distance. And, as one would expect with the derivative, there is a root wherever there is a local minima.

It is important to note however that it is not the actual derivative,

![image](https://github.com/user-attachments/assets/ba96d537-aaa1-41b1-840a-9c7259c375fc)

The above compares $`\sqrt(|p-q(u)|) `$ with that given above. As can be seen the roots are in the same place but the shape differs. Importantly, the sign is inverted. Chen et al propose pruning intervals based on the signs of the interval bounds, as they can be used to ascertain the shape:

![image](https://github.com/user-attachments/assets/3e31ea27-cc5b-4dd8-abe1-d6052988b87c)

However their test is inverted for us since they assert that $` f'(u) = q(u) `$ where $`f(u) = || p-q(u) ||^2`$, which we were unable to demonstrate as shown above.


**Root Finder**

The polynomial is quintic, and so we need an iterative solver of some sort. Most solutions use Sturms theorem, however [Rouillier and Zimmerman](https://inria.hal.science/inria-00072518/file/RR-4113.pdf) propose a compelling method based on Vincent's theorem.

As described in the above, pretty much all methods operate as follows:

1. Bisect the polynomial into intervals
2. Count the roots in each interval
3. If that section has only one root, find the actual root through a naïve search
4. Repeat until all intervals have been found

The "outer loop" of this is the problem of root isolation, which gives a set of intervals containing only one root each. Once we have these intervals, a very simple algorithm - such as bisection comparing the signs of each split - can be used to find the true root within the interval. The simple algorithm can be used safely because there is guaranteed to only be one root, so it's not possible for the algorithm to 'miss' one.

What is particularly interesting to us about Rouillier and Zimmerman's approach is that it works on the principle of transforming polynomials, both to get a polynomial to which Descartes Rule-of-Signs can be applied to count roots, and to get a polynomial representing a specific interval. 

(In this case the intervals are defined by a pair of integers - k & c. These are sort of like rationals, and applying a transformation to the polynomial parameterised by k & c ($` P_{kc}(P(x)) `$), gives a new polynomial that represents just that interval (see [Krandick](https://www.maths.ed.ac.uk/mckinnon/MS/96-021/)). At the end, k & c are resolved to actual interval bounds on the original polynomial for finding the true roots. 

These transformations can be fairly complex. However, they are true 'transformations', in the sense that applying the function to a polynomial results in another polynomial of the same order.

Coupled with the fact we only ever have to deal with one order (5th), we can take advantage of this and use a symbolic maths package (such as Matlabs Symbolic Toolbox) to get each function as a set of basic arithmetic operations applied to the polynomials coefficients, per power. 

For example, define a 5th order polynomial where all the terms are symbolic variables:

```
P(x) =
 a5*x^5 + a4*x^4 + a3*x^3 + a2*x^2 + a1*x + a0
```

We can then write out the equations needed by Rouillier and Zimmerman, for example a Taylor Shift on P:

```
>> T(x) = P(x+1)
T(x) =
a0 + a1*(x + 1) + a2*(x + 1)^2 + a3*(x + 1)^3 + a4*(x + 1)^4 + a5*(x + 1)^5
```

And then have the package rearrange to get the equations for each coefficient:

```
>> collect(T)
ans(x) =
a5*x^5 + (a4 + 5*a5)*x^4 + (a3 + 4*a4 + 10*a5)*x^3 + (a2 + 3*a3 + 6*a4 + 10*a5)*x^2 + (a1 + 2*a2 + 3*a3 + 4*a4 + 5*a5)*x + a0 + a1 + a2 + a3 + a4 + a5
```

We can then just copy the resulting equations into JavaScript and have those functions available in Js, without the need to use any math packages or other imports.

This only works when the parameters of the functions are known however - for example, above we define the Taylor Shift $` T_1(P(x))`$ with a value of 1, but we wouldn't be able to define it for $T_n$.

The approach of Rouillier and Zimmerman however is to search the intervals in such a way as the polynomial representing each interval can be transformed into another adjacent interval in the search, by a small subset of operations.

Specifically, $` H_i(P(x)) = P(xi) `$ and $` HTranslate_i(P(x)) = H_i(T_1(P(x)) `$

As these operations are applied, they change the polynomial to represent an interval over different _k_ & _c_.

For example, transforming $` H_2(P) `$ transforms $` P_{kc} `$ into $` P_{kc'} `$ such that $` k = k + 2 `$ and $` c =  2^2 c `$

For _HTranslate_`, with a Taylor Shift of 1, the transforms are $` k = k + i `$ and $` c = (c + 1)2^k `$

With these, we can make certain steps through the interval, for example to go from an interval defined by k=1, c=1 to k=2,c=2, we'd use $` H_1 `$. To go from k=3,c=1 to k=2,c=1, we'd use $` HTranslate_{-1} `$: because $` c' = (1 + 1) 2^{-1} = 1, and \space k' = k - 1 = 2 `$.

All steps through k & c have to be able to fit into these two functions in such a way, however Rouillier and Zimmerman demonstrate that this is the case when the search is conducted via a depth first traversal of a binary tree.

![image](https://github.com/user-attachments/assets/18266acd-247f-4898-8a6e-b9e7a0394dae)

Fortunately, the scalars of _x_ on _H_ and _HTranslate_ can be parameterised without exploding the complexity of the operations, and HTranslate always has a Taylor Shift of one. The other pre-transform before root counting ($` T_1R(Q) `$ from definition 3 in the paper), can also be found using the symbolic package in the same way as above.

With these in place we can isolate all the roots, and then it is a case of performing a bisection algorithm (such as Algorithm 1 from Chen et al.) to find the true root(s). These represent the points at _u_ on _q_ closest to _p_, so thereafter we can just take the closest point to _p_ (or the end nodes).

**Results**

We can see this now functioning on a number of different curves, including those with cusps & self intersections (note `Num Roots` now applies Chen's pruning). 

![image](https://github.com/user-attachments/assets/47e78886-674c-432d-a69e-1261b4bce30e)

The number of iterations is between 3 and 9 for the found intervals. The total query time, including any initial coefficient calculations that cannot be cached, is between 0 and 0.1 ms for all the above.

**Comment by sebjf** [comment]
An update, line-curve intersections have been added based on substituting the parametric equation of a curve into the implicit equation of a line.

For curve/curve tests, it is more complex than anticipated. The assertion before that curve/curves are quartic only was only for quadratic Beziers, but we deal with cubics.

There are a number of what are known as [clipping algorithms](https://www.sciencedirect.com/science/article/abs/pii/001044859090039F?via%3Dihub) for finding intersections, however some papers claim that for low order curves the root finding method is [still optimal in time](https://www.researchgate.net/publication/220632467_Implicit_representation_of_parametric_curves_and_surfaces). 

The root finding method is similar to the line/curve test, but in this case one of the curves is converted into implicit form, a non-trivial operation, but one for which Sederberg and Goldman in the above say they show is possible.

We would then need to find the roots of a 9th order polynomial. Fortunately, we have a working root finding class. We would need to create versions of the baked methods for 9th order polynomials in addition to 5th order ones (and also not use Chen's filtering) but otherwise the algorithm is identical. We can use the same approach as before to make the baked methods from symbolic expressions.

This seems to be the best approach as is both suggested to be fastest, and uses a lot of existing functionality, whereas other searching algorithms are complex and would need to be written from scratch.

**Comment by sebjf** [comment]
A curve-curve intersection test has been implemented. This is based on Sylvester's method of implicitization, because that was the most straightforward to bake into equations that can be copied into Javascript.

Once this was done the RootFinder was updated to support finding roots of ninth-order polynomials. This was done by moving the transform functions (`H`, `HTranslate` and `countRoots`) into a second class that described the "problem" that RootFinder should work on. The equations for the above functions are determined in the same way (using symbolic libraries, but this time with 9th degree symbolics instead of 5).

Further fixes to the RootFinder had to be made as the bisection algorithm was from Chen which assumes that the functions are moving in a specific direction. Also it appears either the root counting method, or the algorithm in general is not robust to the case where a polynomial contains roots at zero. A condition was added for this.

Finally, the method of curve-curve intersection by root isolation finds t between 0 and 1 on one curve, but that point may be outside the 0 to 1 range on the other.

One way to determine this is to perform curve inversion and find t for a coordinate. There are methods to do this in [Sederberg's book](https://scholarsarchive.byu.edu/cgi/viewcontent.cgi?article=1000&context=facpub#page220). Note that cubics with coincident points must be treated as quadratics, and all colinear curves as lines, if inversion is to be used.

For this reason (the need to handle multiple inversion cases), the quickest to develop solution was used, which is to use the closestPointOnCurve algorithm to check if the intersection point falls on the valid range of the other curve.

The polynomial for the curve-curve intersection cannot be cached in the same way as for the closest point test, because it is not obvious how to delineate terms that only use points on A or B (though it should be possible in theory). The function to build the coefficients is intimidating in the number of operations, however, the intersection test itself runs in between 0 and 0.1 ms - this is surprisingly quick and effectively the same time as the closest point test.

The last test to implement is the bezier self-intersection test, for which there are explicit formulas: https://www.mdpi.com/2227-7390/12/16/2463

**Comment by sebjf** [comment]
The curve-self intersection tests have also been implemented. This is based on the method of [Parcly Taxel](https://math.stackexchange.com/questions/3776840/2d-cubic-bezier-curve-point-of-self-intersection) who shows how by putting the curve in canonical form and using Viete's formulas, the intersection parameter(s) can be found as a simple function of P3 (the 'free' coordinate of the canonical form).

To get the canonical form we use matrix inversion, rather than the [geometric method](https://pomax.github.io/bezierinfo/#canonical), because the geometric method can have singularities.

[Javier Sánchez-Reyes](https://www.mdpi.com/2227-7390/12/16/2463) also has a very nice succinct paper on the explicit formulas, but implementing them does not result in the right answer.

The execution time of this test is trivial compared to the others, and further, the result can be cached in the curve object.

**Comment by sebjf** [comment]
Hey @sanmont3drepo, this is now ready to merge into #4897.

The snapping is based around a class, `SVGSnapHelper`, that loads an SVG from source itself independently of the image component.

In `viewer2D.component.tsx` there is sample code showing how to initialise and use it.

Start by looking here: https://github.com/3drepo/3drepo.io/blob/c9b37dbacd054b1b27f0c3bfbbcbd2fbe43f5fd5/frontend/src/v5/ui/components/viewer/drawingViewer/viewer2D.component.tsx#L181

The `SVGSnapHelper` instance, and the elements for showing the cursor etc for the sample are initialised a little above, here: https://github.com/3drepo/3drepo.io/blob/c9b37dbacd054b1b27f0c3bfbbcbd2fbe43f5fd5/frontend/src/v5/ui/components/viewer/drawingViewer/viewer2D.component.tsx#L94

Let me know if you have any questions!

**Comment by sebjf** [comment]
**Transforms**

The issue with the PDF is that the SVG output by Inkscape includes the use of css-style transform elements. 

These are part of the SVG spec, but we ignored them because ODA doesn't use them, and its more performant if we can just iterate all the elements without having to keep track of the transforms in the tree. However, as a result when they are present they will cause the SVG to render differently to what the snapping controller considers the geometry to be.

Supposedly, Inkscape should be able to bake transforms, but we tried a number of approaches that all failed for the example PDF (failed as in, the export still contained transform attributes).

Supposedly again, the spec method `getCTM()` should give the final transformation matrix of each SVG element - however this only works on [Chrome](https://issues.chromium.org/issues/41334819) if the SVG is attached to the document proper.

It seems then we need to walk the SVG tree and parse the transforms properly.

This is annoying, because now we have a conditional for each entity, along with a matrix transform for each point, whereas before we were just unpacking arrays.

However, the implementation is not difficult since we can use `DOMMatrix` for everything, and there are some additional advantages to the more advanced parsing - such as it no longer considering the `clipPaths` to be snappable.

Adding this stack based transform tracking to `parseSvg` fixes the obvious issues with PDF converted SVGs.


**Alignment**

There is still a problem though, in that the snapped points are slightly offset relative to the rendered geometry.

![image](https://github.com/user-attachments/assets/9a697b2e-6b08-4fff-83be-14fcf6656b2a)

This is almost certainly the result of the SVG image having a viewbox with fractional dimensions. (The exported SVG has the attributes `width="969.03522"` and `height="909.68237"`, for example.)

For SVGs with integer sizes, the alignment is as expected.

The most likely place to introduce this error is the `createImageBitmap` call and its associated logic which gets the difference in projection. `createImageBitmap` only works with integers. The way in which a non-integer _viewBox_ is mapped to a integer canvas is not explicitly documented. 

(It should be noted that the viewerLayer2D is exactly the same size as the background of the projected image.)

If we were to know what `createImageBitmap` does underneath, we could model that in the code that estimates the true projection. However this is an implementation detail and may change between browsers.

The simplest approach is to simply force the `width`, `height` & `viewBox` to be integers, in the same place we set the SVG background colour.

Note that this only has to be done for image that is rendered by `createImageBitmap`. We don't have to change anything for the snapping or transform any primitives. This is because these attributes don't alter the underlying coordinate system of the primitives; rather, they are defined *in* that coordinate system. Rounding fractional coordinates up then will add an additional sub-pixel border to the bottom and the right of the image, but everything relative to the top-left will remain the same. The image will technically have different dimensions to the original file, though imperceptibly so.

**Comment by sebjf** [comment]
Hey @sanmont3drepo, give this another go when you are ready!

**Comment by carmenfan** [comment]
merged into #4897
